File: Fundamentals_of_Software_Design/2_Algorithms/Исследование квадратичной и квази-линейной сортировки/main.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

#define SIZE(arr) (sizeof(arr) / sizeof(arr[0]))

// --- Квадратичные сортировки (O(n^2)) ---

void bubble_sort(int arr[], int n) {
    for (int i = 0; i < n - 1; i++) {
        for (int j = 0; j < n - i - 1; j++) {
            if (arr[j] > arr[j + 1]) {
                int temp = arr[j];
                arr[j] = arr[j + 1];
                arr[j + 1] = temp;
            }
        }
    }
}

void insertion_sort(int arr[], int n) {
    for (int i = 1; i < n; i++) {
        int key = arr[i], j = i - 1;
        while (j >= 0 && arr[j] > key) {
            arr[j + 1] = arr[j];
            j--;
        }
        arr[j + 1] = key;
    }
}

void selection_sort(int arr[], int n) {
    for (int i = 0; i < n - 1; i++) {
        int min_idx = i;
        for (int j = i + 1; j < n; j++) {
            if (arr[j] < arr[min_idx])
                min_idx = j;
        }
        int temp = arr[i];
        arr[i] = arr[min_idx];
        arr[min_idx] = temp;
    }
}

// --- Квази-линейные сортировки (O(n log n)) ---

void quick_sort(int arr[], int low, int high) {
    if (low >= high) return;

    int pivot = arr[high], i = low - 1;
    for (int j = low; j < high; j++) {
        if (arr[j] < pivot) {
            i++;
            int temp = arr[i];
            arr[i] = arr[j];
            arr[j] = temp;
        }
    }
    int temp = arr[i + 1];
    arr[i + 1] = arr[high];
    arr[high] = temp;

    int pi = i + 1;
    quick_sort(arr, low, pi - 1);
    quick_sort(arr, pi + 1, high);
}

void merge(int arr[], int left, int mid, int right) {
    int n1 = mid - left + 1, n2 = right - mid;
    int L[n1], R[n2];

    for (int i = 0; i < n1; i++) L[i] = arr[left + i];
    for (int i = 0; i < n2; i++) R[i] = arr[mid + 1 + i];

    int i = 0, j = 0, k = left;
    while (i < n1 && j < n2) {
        if (L[i] <= R[j]) arr[k++] = L[i++];
        else arr[k++] = R[j++];
    }
    while (i < n1) arr[k++] = L[i++];
    while (j < n2) arr[k++] = R[j++];
}

void merge_sort(int arr[], int left, int right) {
    if (left >= right) return;
    int mid = left + (right - left) / 2;
    merge_sort(arr, left, mid);
    merge_sort(arr, mid + 1, right);
    merge(arr, left, mid, right);
}

void heapify(int arr[], int n, int i) {
    int largest = i, left = 2 * i + 1, right = 2 * i + 2;
    if (left < n && arr[left] > arr[largest]) largest = left;
    if (right < n && arr[right] > arr[largest]) largest = right;
    if (largest != i) {
        int temp = arr[i];
        arr[i] = arr[largest];
        arr[largest] = temp;
        heapify(arr, n, largest);
    }
}

void heap_sort(int arr[], int n) {
    for (int i = n / 2 - 1; i >= 0; i--) heapify(arr, n, i);
    for (int i = n - 1; i > 0; i--) {
        int temp = arr[0];
        arr[0] = arr[i];
        arr[i] = temp;
        heapify(arr, i, 0);
    }
}

// --- Функции для тестирования и измерения времени ---

void copy_array(int src[], int dest[], int n) {
    for (int i = 0; i < n; i++) dest[i] = src[i];
}

void test_sorts() {
    int test_cases[][10] = {
        {},
        {42},
        {5, 2, 8, 3, 1},
        {3, 3, 3, 3},
        {-2, -5, 0, 7, 1},
        {100, 99, 98, 97, 96}
    };
    int n_cases = SIZE(test_cases);

    void (*sorts[])(int[], int) = {bubble_sort, insertion_sort, selection_sort, heap_sort};
    char *names[] = {"Bubble", "Insertion", "Selection", "Heap"};

    for (int i = 0; i < 4; i++) {
        for (int j = 0; j < n_cases; j++) {
            int arr[10];
            copy_array(test_cases[j], arr, 10);
            sorts[i](arr, 10);
            for (int k = 1; k < 10; k++) {
                if (arr[k - 1] > arr[k]) {
                    printf("Ошибка в %s\n", names[i]);
                    return;
                }
            }
        }
    }
    printf("Все тесты пройдены!\n");
}

void benchmark() {
    int sizes[] = {100, 500, 1000, 5000, 10000, 20000};
    int num_sizes = SIZE(sizes);
    void (*sorts[])(int[], int) = {bubble_sort, insertion_sort, selection_sort, heap_sort};
    char *names[] = {"Bubble", "Insertion", "Selection", "Heap"};

    printf("%-10s %-10s %-10s %-10s %-10s\n", "Размер", "Bubble", "Insert", "Select", "Heap");

    for (int i = 0; i < num_sizes; i++) {
        int size = sizes[i];
        int arr[size], copy[size];
        
        for (int j = 0; j < size; j++) arr[j] = rand() % 10000;

        printf("%-10d", size);
        for (int j = 0; j < 4; j++) {
            copy_array(arr, copy, size);
            clock_t start = clock();
            sorts[j](copy, size);
            clock_t end = clock();
            printf("%-10.4f", (double)(end - start) / CLOCKS_PER_SEC);
        }
        printf("\n");
    }
}

// --- Главная функция ---
int main() {
    srand(time(NULL));
    test_sorts();
    benchmark();
    return 0;
}

File: Fundamentals_of_Software_Design/2_Algorithms/Исследование квадратичной и квази-линейной сортировки/main.py
import time
import random
from typing import List

# --- Квадратичные сортировки ---
def bubble_sort(arr: List[int]) -> List[int]:
    """Сортировка пузырьком."""
    n = len(arr)
    for i in range(n):
        for j in range(n - 1 - i):
            if arr[j] > arr[j + 1]:
                arr[j], arr[j + 1] = arr[j + 1], arr[j]
    return arr

def insertion_sort(arr: List[int]) -> List[int]:
    """Сортировка вставками."""
    for i in range(1, len(arr)):
        key = arr[i]
        j = i - 1
        while j >= 0 and arr[j] > key:
            arr[j + 1] = arr[j]
            j -= 1
        arr[j + 1] = key
    return arr

def selection_sort(arr: List[int]) -> List[int]:
    """Сортировка выбором."""
    n = len(arr)
    for i in range(n):
        min_idx = i
        for j in range(i + 1, n):
            if arr[j] < arr[min_idx]:
                min_idx = j
        arr[i], arr[min_idx] = arr[min_idx], arr[i]
    return arr

# --- Квази-линейные сортировки ---
def quick_sort(arr: List[int]) -> List[int]:
    """Быстрая сортировка."""
    if len(arr) <= 1:
        return arr
    pivot = arr[len(arr) // 2]
    left = [x for x in arr if x < pivot]
    middle = [x for x in arr if x == pivot]
    right = [x for x in arr if x > pivot]
    return quick_sort(left) + middle + quick_sort(right)

def merge_sort(arr: List[int]) -> List[int]:
    """Сортировка слиянием."""
    if len(arr) <= 1:
        return arr
    mid = len(arr) // 2
    left = merge_sort(arr[:mid])
    right = merge_sort(arr[mid:])
    return merge(left, right)

def merge(left: List[int], right: List[int]) -> List[int]:
    """Слияние массивов для merge_sort."""
    result = []
    i = j = 0
    while i < len(left) and j < len(right):
        if left[i] < right[j]:
            result.append(left[i])
            i += 1
        else:
            result.append(right[j])
            j += 1
    result.extend(left[i:])
    result.extend(right[j:])
    return result

def heap_sort(arr: List[int]) -> List[int]:
    """Пирамидальная сортировка (HeapSort)."""
    def heapify(arr, n, i):
        largest = i
        l = 2 * i + 1
        r = 2 * i + 2
        if l < n and arr[l] > arr[largest]:
            largest = l
        if r < n and arr[r] > arr[largest]:
            largest = r
        if largest != i:
            arr[i], arr[largest] = arr[largest], arr[i]
            heapify(arr, n, largest)

    n = len(arr)
    for i in range(n // 2 - 1, -1, -1):
        heapify(arr, n, i)
    for i in range(n - 1, 0, -1):
        arr[i], arr[0] = arr[0], arr[i]
        heapify(arr, i, 0)
    return arr

# Функция измерения времени работы алгоритма
def measure_time(sort_func, arr: List[int]) -> float:
    start = time.perf_counter()
    sort_func(arr[:])  # Передаем копию массива, чтобы не изменять оригинал
    end = time.perf_counter()
    return end - start

# Тестирование корректности
def test_sorting():
    test_cases = [
        [],
        [42],
        [5, 2, 8, 3, 1],
        [3, 3, 3, 3],
        [-2, -5, 0, 7, 1],
        list(range(100, 0, -1))
    ]
    sorts = [bubble_sort, insertion_sort, selection_sort, quick_sort, merge_sort, heap_sort]
    
    for sort_func in sorts:
        for arr in test_cases:
            assert sort_func(arr[:]) == sorted(arr)
    
    print("Все тесты пройдены!")

# Запуск эксперимента
def benchmark():
    sizes = [100, 200, 400, 800, 1600, 3200]  # Размеры массивов
    sorts = {
        "Bubble Sort": bubble_sort,
        "Insertion Sort": insertion_sort,
        "Selection Sort": selection_sort,
        "Quick Sort": quick_sort,
        "Merge Sort": merge_sort,
        "Heap Sort": heap_sort
    }
    
    print(f"{'Размер':<8} {'Bubble':<10} {'Insert':<10} {'Select':<10} {'Quick':<10} {'Merge':<10} {'Heap':<10}")
    
    for size in sizes:
        arr = [random.randint(-10000, 10000) for _ in range(size)]
        times = {name: measure_time(func, arr) for name, func in sorts.items()}
        
        print(f"{size:<8} {times['Bubble Sort']:.4f} {times['Insertion Sort']:.4f} {times['Selection Sort']:.4f} {times['Quick Sort']:.4f} {times['Merge Sort']:.4f} {times['Heap Sort']:.4f}")

if __name__ == "__main__":
    test_sorting()
    benchmark()
